### PR TITLE
Fix grainy donut chart MWPW-116408

### PR DESCRIPTION
**Pull request by Brandon32**
* Fix grainy donut chart labels

Resolves: [MWPW-116408](https://jira.corp.adobe.com/browse/MWPW-116408)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/bmarshal/chartsqa?martech=off
- After: https://grainy-donut--milo--adobecom.hlx.page/drafts/bmarshal/chartsqa?martech=off

### CODE REVIEW

**Review by fullcolorcoder — APPROVED** [review]

**Review by sanjayms01 — APPROVED** [review]

File: libs/blocks/chart/chart.js [hunk 1]
```diff
@@ -405,19 +406,45 @@ export const getChartOptions = (chartType, data, colors, size, chart) => {
...
+};
+
 const initChart = (chartWrapper, chartType, data, colors, size) => {
   const themeName = getTheme(size);
   const chart = window.echarts?.init(chartWrapper, themeName, { renderer: 'svg' });
-  const chartOptions = getChartOptions(chartType, data, colors, size, chart);
+  const chartOptions = getChartOptions(chartType, data, colors, size);
 
   chart.setOption(chartOptions);
+  setListeners(chart, chartType, data);
```
[the remark this answers is not in the document]
Comment (Brandon32) [reply libs/blocks/chart/chart.js:447]: I had to move the listeners now that they had to be outside of `donutSeriesOptions`. I tried this a couple way but was running into the issue that `getOptions` and `donutTitleOptions` was setting up the chart, coupling / side effects. There is some duplicate code but it makes it clearer what's happening inside of `initChart` and that duplicate code could be refactored and some variables can be removed.

[the remark this answers is not in the document]
Comment (meganthecoder) [reply libs/blocks/chart/chart.js:447]: Ok, if you need to keep it here, then I think reducing duplicate code should be done now not later. Maybe add more data processing to `initChart` and pass them in to both `getChartOptions` and `setListeners`.

[the remark this answers is not in the document]
Comment (meganthecoder) [reply libs/blocks/chart/chart.js:447]: The conditional should go around `setListeners` in `initChart`. If it's only for the donut chart, there's no reason to waste processing for other chart types. Also, I don't see a reason for `setDonutListeners` and `setListeners` to be separate. Seems like you should combined them

[the remark this answers is not in the document]
Comment (Brandon32) [reply libs/blocks/chart/chart.js]: Updated, please take a look at `initChart` and `getChartOptions`

File: libs/blocks/chart/chart.js [hunk 2]
```diff
@@ -404,22 +407,40 @@ export const getChartOptions = (chartType, data, colors, size, chart) => {
...
+  // Remove header names
+  const sourceData = (source && source[0].every((i) => typeof i === 'string')) ? source.slice(1) : source;
+  const sum = sourceData?.reduce((total, current) => total + current[0], 0);
+  const firstSeries = seriesData?.[0];
+  const title = firstSeries ? propertyValueCI(firstSeries, 'title') : '';
+  let mouseOutValue = sum;
+
+  chart.on('mouseover', (value) => setDonutTitle(chart, value?.data?.[0], units?.[0], title));
+  chart.on('mouseout', () => setDonutTitle(chart, mouseOutValue, units?.[0], title));
+  chart.on('legendselectchanged', ({ selected }) => { mouseOutValue = handleDonutSelect(sourceData, selected, chart, units?.[0], title); });
```
Comment (meganthecoder) [inline libs/blocks/chart/chart.js:429]: Nit: add `units?.[0]` to a variable since it's used more than once

**Review by meganthecoder — APPROVED** [review]
Looks good! Just one more small suggestion

File: libs/blocks/chart/chart.js [hunk 3]
```diff
@@ -330,18 +333,17 @@ export const pieSeriesOptions = (size) => {
...
+ * @param {object} series
+ * @param {array} headers
  * @param {array} colors
  * @param {string} size
+ * @param {array} units
  * @returns {object}
  */
-export const getChartOptions = (chartType, data, colors, size, chart) => {
-  const headers = data?.data?.[0];
+export const getChartOptions = (chartType, dataset, series, headers, colors, size, units = []) => {
```
Comment (chrischrischris) [inline libs/blocks/chart/chart.js:344]: don't have to change but this would be a good candidate for object param instead

**Pull request merged**